File: src/leetcode/medium/add-two-numbers-2/js/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  var preHead = new ListNode(0);
  var p = preHead;
  var extra = 0;

  while (l1 || l2 || extra) {
    var sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + extra;
    extra = Math.floor(sum / 10);

    p.next = new ListNode(sum % 10);
    p = p.next;

    l1 = l1 ? l1.next : l1;
    l2 = l2 ? l2.next : l2;
  }

  return preHead.next;
};

File: src/leetcode/medium/add-two-numbers-2/c/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
  struct ListNode preHead;
  struct ListNode* p = &preHead;
  int sum = 0;

  while (l1 != NULL || l2 != NULL || sum != 0) {
    if (l1 != NULL) {
      sum += l1->val;
      l1 = l1->next;
    }

    if (l2 != NULL) {
      sum += l2->val;
      l2 = l2->next;
    }

    p = p->next = (struct ListNode*)malloc(sizeof(struct ListNode));
    p->val = sum % 10;

    sum /= 10;
  }
  p->next = NULL;

  return preHead.next;
}
